Infrastructure-Scripts/Set-AppRoleAssignments/Set-AppRoleAssignments.ps1

```diff
@@ -0,0 +1,142 @@
+        }
+        elseif ($ServicePrincipal.Count -gt 1) {
+            throw "-> Found a duplicate app registrations with the same display name. Investigate"
+        }
+        else {
+            Write-Output "-> App registration $AppRegistrationName not found in AAD - Creating"
+
+            if (!$DryRun) {
+                New-AppRegistration -AppRegistrationName $AppRegistrationName -IdentifierUri $IdentifierUri
+            }
+
+            Write-Output "  -> Successfully created app registration - $AppRegistrationName"
+            $ServicePrincipal = Get-ServicePrincipal -DisplayName $AppRegistrationName
+        }
+
+        foreach ($AppRole in $AppRegistration.appRoles) {
+            $MatchedAppRole = $ServicePrincipal.appRoles | Where-Object { $_.value -eq $AppRole.appRoleName }
+
+            if ($MatchedAppRole) {
+                Write-Output "  -> Processing app role $($AppRole.appRoleName)"
+            }
+            else {
+                Write-Output "  -> App role $($AppRole.appRoleName) not found on app registration - Creating App Role"
+
+                if (!$DryRun) {
+                    New-AppRegistrationAppRole -AppRoleName $AppRole.appRoleName -IdentifierUri $IdentifierUri
+                }
+
+                Write-Output "    -> Successfully added app role $($AppRole.appRoleName)"
+
+                $ServicePrincipal = Get-ServicePrincipal -DisplayName $AppRegistrationName
+                $MatchedAppRole = $ServicePrincipal.appRoles | Where-Object { $_.value -eq $AppRole.appRoleName }
+            }
+            foreach ($MIResource in $AppRole.resourceNameSuffix) {
+                $ManagedIdentityResourceName = "$ResourceNamePrefix$MIResource"
+                $ManagedIdentity = Get-ServicePrincipal -DisplayName $ManagedIdentityResourceName
+                if ($ManagedIdentity.Count -eq 1) {
+                    Write-Output "    -> Processing Managed Identity of $ManagedIdentityResourceName"
+                }
+                elseif ($ManagedIdentity.Count -gt 1) {
+                    Write-Output "    -> Found duplicate app registrations with name $ManagedIdentityResourceName - Investigate"
+                    continue
+                }
+                else {
+                    Write-Output "    -> Managed Identity of $ManagedIdentityResourceName not found"
+                    continue
+                }
+
+                try {
+                    if ($ServicePrincipal.ObjectId) {
+                        $AppRoleAssignments = Get-AppRoleAssignments -ServicePrincipalId $ServicePrincipal.ObjectId
+                        $AppRoleAssignmentExists = $AppRoleAssignments.value | Where-Object { $_.appRoleId -eq $MatchedAppRole.id -and $_.principalId -eq $ManagedIdentity.ObjectId }
+                    }
+                    else {
+                        $AppRoleAssignmentExists = $null
+                    }
+
+                    if ($AppRoleAssignmentExists) {
```

Comment:
Could this block go to line 114? Wouldn't need the else and makes it easier to follow code branches